aggregate/billing_functions/utils.py

```diff
@@ -0,0 +1,742 @@
+        'memory/n1-preemptible/1': 'mib * msec',
+        'service-fee/1': '$/msec',
+    }.get(batch_resource_type, batch_resource_type)
+
+
+def get_start_and_end_from_request(
+    request,
+) -> Tuple[Optional[datetime], Optional[datetime]]:
+    """
+    Get the start and end times from the cloud function request.
+    """
+    if request:
+        return request.params['start'], request.params['end']
+    return (None, None)
+
+
+def date_range_iterator(
+    start,
+    end,
+    intv=DEFAULT_RANGE_INTERVAL,
+) -> Iterator[Tuple[datetime, datetime]]:
+    """
+    Iterate over a range of dates.
+    """
+    dt_from = start
+    dt_to = start + intv
+    while dt_to < end:
+        yield (dt_from, dt_to)
+        dt_from += intv
+        dt_to += intv
+
+    dt_to = min(dt_to, end)
+    if dt_from < end:
+        yield (dt_from, end)
+
+
+def get_start_and_end_from_data(data) -> Tuple[Optional[datetime], Optional[datetime]]:
+    """
+    Get the start and end times from the cloud function data.
+    """
+    if data:
+        dates = {}
+        if data.get('attributes'):
+            dates = data.get('attributes', {})
+        elif data.get('start') and data.get('end'):
+            dates = data
+        elif data.get('message'):
+            dates = dict(json.loads(data['message'])) or {}
+
+        logger.info(f'data: {data}, dates: {dates}')
+
+        start = dates.get('start', '')
+        end = dates.get('end', '')
+
+        try:
+            start, end = datetime.fromisoformat(start), datetime.fromisoformat(end)
+        except ValueError as error:
+            logger.error(f'Error: {error}')
+            return (None, None)
+
+        return (start, end)
+
+    return (None, None)
+
+
+def process_default_start_and_end(start, end) -> Tuple[datetime, datetime]:
+    """
+    Process the start and end times.
+    """
+    if not end:
+        # start of today
+        end = datetime.now().replace(hour=0, minute=0, second=0, microsecond=0)
+
+    if not start:
+        start = end - DEFAULT_RANGE_INTERVAL
+
+    assert isinstance(start, datetime) and isinstance(end, datetime)
+    return start, end
+
+
+def get_date_intervals_for(
+    start, end, interval=DEFAULT_RANGE_INTERVAL
+) -> Iterator[Tuple[datetime, datetime]]:
+    """Process start and end times (and get defaults)"""
+    s, e = process_default_start_and_end(start, end)
+    return date_range_iterator(s, e, intv=interval)
+
+
+def read_secret(project_id: str, secret_name: str) -> Optional[str]:
+    """Reads the latest version of a GCP Secret Manager secret.
+
+    Returns None if the secret doesn't exist."""
+
+    secret_manager = secretmanager.SecretManagerServiceClient()
+    secret_path = secret_manager.secret_path(project_id, secret_name)
+
+    response = secret_manager.access_secret_version(
+        request={'name': f'{secret_path}/versions/latest'}
+    )
+    return response.payload.data.decode('UTF-8')
+
+
+def get_entry(
```

Comment:
Clarify this is get_entry for something related to hail, not a GCP line item eg

Reply:
Renamed to `get_hail_entry`